pydantic/main.py

```diff
@@ -592,6 +603,51 @@ def __repr_args__(self) -> _repr.ReprArgs:
             if not k.startswith('_') and (k not in self.model_fields or self.model_fields[k].repr)
         ]
 
+    def __class_getitem__(cls, item: typing.Union[type[Any], typing.Tuple[type[Any], ...]]) -> type[Any]:
+        def _cache_key(_params: Any) -> typing.Tuple[type[Any], Any, typing.Tuple[Any, ...]]:
+            return cls, _params, typing_extensions.get_args(_params)
+
+        cached = _generic_types_cache.get(_cache_key(item))
+        if cached is not None:
+            return cached
+
+        if not isinstance(item, tuple):
+            item = (item,)
+
+        if not is_generic_model(cls):
+            raise TypeError('Cannot parameterize a non-generic BaseModel subclass')
+
+        generic_alias = super().__class_getitem__(item)  # type: ignore[misc]
+
+        model_name = cls.__concrete_name__(item)
+        submodel = create_generic_submodel(model_name, cls)
+
+        typevars_map = dict(zip(submodel.__parameters__, item))
+        submodel.model_rebuild(force=True, typevars_map=typevars_map)
+        submodel.__parameters__ = generic_alias.__parameters__
+
+        _generic_types_cache[_cache_key(item)] = submodel
+        if len(item) == 1:
+            _generic_types_cache[_cache_key(item[0])] = submodel
+
+        return submodel
+
+    @classmethod
+    def __concrete_name__(cls: type[Any], params: typing.Tuple[type[Any], ...]) -> str:
+        """Compute class name for child classes.
+
+        :param params: Tuple of types the class . Given a generic class
+            `Model` with 2 type variables and a concrete model `Model[str, int]`,
+            the value `(str, int)` would be passed to `params`.
+        :return: String representing a the new class where `params` are
+            passed to `cls` as type variables.
+
+        This method can be overridden to achieve a custom naming scheme for GenericModels.
+        """
+        param_names = [_repr.display_as_type(param) for param in params]
```

Comment:
if this is only used on generics, should we raise an error if the model is not generic?

Reply:
I'm okay either way; I can't really imagine a circumstance where you'd want to use it for non-generic-models, if someone found a use case where it was worth using on non-generic models for some reason, they could always override the method to drop the check for generic and open an issue requesting it work for non-generic models..